// parcel-namer-rewrite/lib/Config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Config = void 0;

var _logger = require("@parcel/logger");

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const PACKAGE_JSON_SECTION = "parcel-namer-rewrite";

class Config {
  /**
   * Disable namer in development
   */

  /**
   * Disable name hashing in development
   */

  /**
   * Disable logging names
   */

  /**
   * Use file name hashes from parcel
   */
  constructor() {
    _defineProperty(this, "rules", void 0);

    _defineProperty(this, "chain", void 0);

    _defineProperty(this, "developmentDisable", false);

    _defineProperty(this, "developmentHashing", false);

    _defineProperty(this, "silent", false);

    _defineProperty(this, "useParcelHash", true);

    this.chain = '@parcel/namer-default';
    this.rules = [];
  }

  loadFromPackageFolder(rootFolder, logger) {
    const packageJson = _fs.default.readFileSync(_path.default.join(rootFolder, 'package.json')).toString();

    const packageInfo = JSON.parse(packageJson);
    const packageSection = packageInfo[PACKAGE_JSON_SECTION];

    if (!packageSection) {
      logger.warn({
        message: `no "${PACKAGE_JSON_SECTION}" section in package.json. Use no-rules config`
      });
      return;
    }

    if (packageSection && 'chain' in packageSection) {
      this.chain = packageSection.chain;
    }

    this.silent = packageSection && 'silent' in packageSection && packageSection.silent;

    if (packageSection && 'useParcelHash' in packageSection) {
      this.useParcelHash = !!packageSection.useParcelHash;
    }

    if (packageSection && 'rules' in packageSection) {
      Object.keys(packageSection.rules).forEach(k => {
        const ruleData = packageSection.rules[k];
        const ruleTo = typeof ruleData === 'string' ? ruleData : null;

        if (ruleTo === null) {
          logger.warn(`No "to" rule for test "${k}" `);
          return;
        }

        this.rules.push({
          test: new RegExp(k),
          to: ruleTo
        });
      });
    }

    if (packageSection && 'developmentHashing' in packageSection) {
      this.developmentHashing = !!packageSection.developmentHashing;
    }

    if (packageSection && 'developmentDisable' in packageSection) {
      this.developmentDisable = !!packageSection.developmentDisable;
    }
  }

  selectRule(name) {
    const matches = this.rules.map(rule => rule.test.test(name) ? rule : null).filter(rule => rule != null);

    if (matches.length > 0) {
      return matches[0];
    }

    return null;
  }

}

exports.Config = Config;
